Caching a compiled version onSave

mitkonikov opened the pull request:
## Overview

If the user wants to compile and cache a version on each save command, he can do so. This allows them to run the test cases instantly without waiting for it to compile.

## Changes
 - Each problem has one more optional variable `skipNextCompile` which is set to true if the problem is compiled *onSave*
 - When the problem is compiled *onSave* it synchronizes the *JudgeView*
    - Here, we can furthermore improve it by adding an icon in the JudgeView *(I didn't do it)*
 - In the preferences, the user can set this option (compileOnSave) to true, **by default it's false and won't affect anything**.
 - Created `shouldCompile` function which checks if the user wants to `compileOnSave` and if the problem has been already compiled

## How to use it?
 - Enable it in the settings - `compileOnSave`
 - On each save, it will compile.
 - Just click `Run All` or `Run Testcases` and the extension will run the test cases immediately without the need of compiling.
 > Note: The extension will compile the test cases again if the document is `dirty` - means something is changed and not saved.

### Notes
 - The current behaviour is that when a problem is compiled onSave and then the test cases are executed, the variable `skipNextCompile` is reset to false, indicating that the next time the user tries to run the test cases, no matter if it's cached, it will be compiled once again.
 - i.e. the binary is deleted upon each run (as before, no changes there)

### Additional minor fixes:
 - Fixed a comment in the `.vscode/launch.json` - the preLaunchTask was commented out.
 - The command for testing is `npm run test`, not `npm run jest`

### TODO
 - When the pull request is accepted, update the documentation and guide accordingly.

mitkonikov committed:
Caching a compiled version onSave

If the user wants to compile and cache a version on each save command, he can do so. This allows then to run the testcases instantly without waiting it to compile.

agrawal-d commented:
Hi @mitkonikov , this feature looks useful! I'll take a look at it and get back to you.

mitkonikov committed:
Fix a bug where the UI was not properly synced

If the cached version is executed, then the skipNextCompile variable should be reset back to false even in the UI.

agrawal-d commented on .prettierrc.json:
```diff
@@ -4,5 +4,6 @@
     "trailingComma": "all",
-    "proseWrap": "always"
+    "proseWrap": "always",
+    "endOfLine": "auto"
```
What's the reason for these changes? If it's unintentional, please revert them.

mitkonikov replied:
ES Lint and Prettier were giving me errors on each new line. 
In the image, when I remove `endOfLine: "auto"`, you can see errors on each line because eslint thinks there is some other character. I'm using Windows tho, and if this is developed on Linux, it could be due to that.

![image](https://user-images.githubusercontent.com/6832136/138548793-927ee903-fec7-43c6-b273-243c854d95c4.png)

agrawal-d replied:
It's certainly due to Windows. You'll need to configure VS Code to use LF endings instead of CRLF.

mitkonikov replied:
So, do I revert this change and add inside the documentation that Windows Devs have to change the Line Endings or leave it for other Windows Devs by default?

agrawal-d replied:
Yeah please revert this change, and yes, adding instructions for Windows will be helpful for other developers.

agrawal-d commented on src/extension.ts:
```diff
@@ -8,7 +8,8 @@ import {
 import { submitToCodeForces, submitToKattis } from './submit';
 import JudgeViewProvider from './webview/JudgeView';
-import { getRetainWebviewContextPref } from './preferences';
+import { getCompileOnSavePref, getRetainWebviewContextPref } from './preferences';
```
Please fix this prettier CI issue.

agrawal-d commented on src/judge.ts:
```diff
@@ -11,7 +11,7 @@ export const isResultCorrect = (
 ): boolean => {
     const expectedLines = testCase.output.trim().split('\n');
-    const resultLines = stdout.trim().split(EOL);
+    const resultLines = stdout.trim().split(EOL).join('\n').split('\n');
```
What's the reason for this change?

mitkonikov replied:
On Windows, the EOL character is "\r\n" which when given to the split function, the split function interprets that both the characters must exist together in a continuation. On the image, I first `require("os")` and show you that the result is not split at all on Windows, hence failing the tests.

![image](https://user-images.githubusercontent.com/6832136/138549191-cee9876b-256e-4fff-bbb2-850c352b2350.png)

agrawal-d commented on src/types.ts:
```diff
@@ -51,6 +52,7 @@ export type Problem = {
     tests: TestCase[];
     srcPath: string;
     local?: boolean;
+    skipNextCompile?: boolean;
```
Lets make the property `skipNextCompile` mandatory. You'll need to initialize this in `handleNewProblem()` in `companion.ts` when we process a problem from Competitive Companion.

agrawal-d reviewed (changes requested):
The code looks mostly good ( apart from the bugs ). I've left some review comments, please take a look.
There are two bugs:

1. Cached compile is always being used when there are no changes.

![bug1](https://user-images.githubusercontent.com/7714968/138548530-0a5701e4-9c62-48ee-9e20-88a95141b2b6.gif)


2. When the editor is dirty and I run a testcase, the result is always blank for the first time.
![bug2](https://user-images.githubusercontent.com/7714968/138548555-31c7d38a-c850-41e4-8829-f1ee9a88518c.gif)

agrawal-d commented:
There's another issue that needs to be fixed. When the document is saved, you send an new problem to the UI. This results in the UI losing it's state ( past run results etc.) Instead of sending a new problem, send just the `skipNextCompile` value using a new `VSToWebViewMessage` command and handle it in React in `App.tsx`.

agrawal-d commented:
Also the terminology `skipNextCompile` is confusing as we have a similar naming to skip compilation for interpreted languages ( Python ). Instead, please use variable names like `useCachedCompilation` and similar words for this logic.

mitkonikov commented:
> The code looks mostly good ( apart from the bugs ). I've left some review comments, please take a look. There are two bugs:
> 
> 1. Cached compile is always being used when there are no changes.
> 2. When the editor is dirty and I run a test case, the result is always blank for the first time.

I will try to reproduce both of the issues and make the appropriate requested changes.
**Thank you so much for looking at the feature, it's an honour to make it, because if made well, it should save time.**

Don't expect quick changes due to how busy I'm at the moment and the work required, but in a day or two, we'll have commits for sure 😅

mitkonikov committed:
Revert prettierrc.json, set LF as default line ending

Tell VSCode that LF should be the default line ending even on Windows.

mitkonikov commented:
![image](https://user-images.githubusercontent.com/6832136/138684095-6a7a9652-f214-4dab-8c2b-94404605b641.png)

I think that the problem is that when we call `textEditor.save()` or `editor.save()`, the `vscode.workspace.onDidSaveTextDocument` event handler is triggered resulting in calling the function `compileOnSave` which compiles the file again, setting the `useCachedCompilation` back to true, even tho we haven't personally saved the document.

I fixed the other things, renamed the variable to `useCachedCompilation`, made it mandatory, but one of the bugs still persists. I'm trying to figure out fixes. Also, I added a caching icon on the UI, so the user will know when there's a cached version.

I can commit the changes till now, but first I wanted to ask you, do you see a workaround for this problem, and how?
> Now, this might not be the only problem or I could be wrong, but I'm still trying to figure out what is happening because the synchronization is complex. It could be the case that React is not propagating the update well. Because the last 'save' command contains the problem with the `useCacheCompilation: true`.

![image](https://user-images.githubusercontent.com/6832136/138685113-34e65d41-6205-415e-af1d-ea2ea37e7ad2.png)
Also, I now update the problem with a different message to the UI.

---

![image](https://user-images.githubusercontent.com/6832136/138688302-94779498-a14e-4a8d-8aef-f843749b30cb.png)
The problem itself ❓

mitkonikov commented:
This pull request will be closed, we combined the work in another branch and the latest pull request about this feature is #212
